towhee/dag/graph_repr.py

```diff
@@ -46,12 +44,118 @@ def operators(self) -> Dict[str, OperatorRepr]:
     def dataframes(self) -> Dict[str, DataframeRepr]:
         return self._dataframes
 
+    @staticmethod
+    def dfs(cur: str, adj: Dict[str, List[str]], flag: Dict[str, int], cur_list: List[str]) -> Tuple[bool, str]:
+        """Depth-First Search the graph
+        Args:
+            cur(`str`):
+                The name of current dataframe.
+            adj(`dict`):
+                A dict store the adjacent dataframe of each dataframe.
+            flag(`dict`):
+                A dict store the status of the columns.
+                - 0 means the dataframe has not been visted yet.
+                - 1 means the dataframe has been visted in this search, which means this
+                columns is part of a loop.
+                - 2 means the dataframe has been searched and confirmed not to be a part
+                of a loop.
+            cur_list('list):
+                The list of dataframe that have been visited in this search.
+        Returns:
+            Return `False` if there is no loop, else `True` and the loop message.
+        """
+        # if `cur` is not the input of any operator, it will not form a loop anyway
+        if cur not in flag:
+            return False, ''
+        # If `cur` has been searched and not in a loop
+        if flag[cur] == 2:
+            return False, ''
+        # If `cur` has been visited in this search
+        if flag[cur] == 1:
+            return True, f'The dataframes {cur_list} forms a loop.'
+        flag[cur] = 1
+
+        for col in adj[cur]:
+            cur_list.append(cur)
+            status, msg = GraphRepr.dfs(col, adj, flag, cur_list)
+            if status:
+                return status, msg
+        flag[cur] = 2
+
+        return False, ''
+
+    def has_loop(self) -> Tuple[bool, str]:
+        """Check if there are loop(s) inside the graph.
+        Returns:
+            Return `False` if there is no loop, else `True` and the loop message.
+        """
+        adj = {}
+        flag = {}
+        dataframes = []
+
+        for op in self._operators:
+            out = op.outputs[0]['df']
+            for df in op.inputs:
+                name = df['df']
+                if name not in dataframes:
+                    adj[name] = []
+                    flag[name] = 0
+                    dataframes.append(name)
+                adj[name].append(out)
+
+        # Recursive DFS
+        for df in dataframes:
+            status, msg = GraphRepr.dfs(df, adj, flag, [])
+            if status:
+                return status, msg
+
+        return False, ''
+
+    def has_isolated_df(self):
+        iso_df = set()
+
+        for df in self._dataframes:
+            iso_df.add(df.name)
```

Comment:
You can change this into a set comprehension (more Pythonic):
`iso_df = {df.name for df in self._dataframes}`